### PR TITLE
Perform more throughout ChassisConfig validation on Stratum-bf

### PR DESCRIPTION
**Pull request by pudelkoM**
We're not validating some mandatory aspects of the ChassisConfig, e.g. SingletonPorts or the Platform value. This PR ports over some of the checks from the bcm chassis manager.

Some of this code is shared with `BcmChassisManager` and could probably be placed in `ConfigMonitoringService::VerifyChassisConfig`.

- [x] <s>Fix port IDs in Tofino default configs (can't be zero)</s> Disabled until we decouple port IDs from SDN ports

### CODE REVIEW

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 1]
```diff
@@ -359,6 +360,67 @@ ::util::Status BFChassisManager::VerifyChassisConfig(
...
+    }
+  }
+
+  // Go over all the singleton ports in the config:
+  // 1- Validate the basic singleton port properties.
+  // 2- Make sure there is no two ports with the same (slot, port, channel).
+  // 3- Make sure all the ports with the same (slot, port) have the same
+  //    speed.
+  // 4- Make sure for each (slot, port) pair, the channels of all the ports
+  //    are valid. This depends on the port speed.
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:396]: Not checked.

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 2]
```diff
@@ -359,6 +360,67 @@ ::util::Status BFChassisManager::VerifyChassisConfig(
...
+  // 5- Make sure no singleton port has the reserved CPU port ID. CPU port is
+  //    a special port and is not in the list of singleton ports. It is
+  //    configured separately.
+  // 6- Make sure IDs of the singleton ports are unique per node.
+  // 7- Make sure IDs match the SDE dev port numbers, as there is not way to do
+  //    port translation on Tofino at the moment.
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:402]: Remove

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 3]
```diff
@@ -359,6 +360,67 @@ ::util::Status BFChassisManager::VerifyChassisConfig(
     return MAKE_ERROR(ERR_INVALID_PARAM)
            << "Port groups are not supported on Tofino.";
   }
+  if (config.nodes_size() != 1) {
+    return MAKE_ERROR(ERR_INVALID_PARAM)
+           << "Multiple nodes are not supported on Tofino.";
+  }
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:366]:
```suggestion
  CHECK_RETURN_IF_FALSE(config.nodes_size())
         << "Multiple nodes are not supported on Tofino.";
```

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 4]
```diff
@@ -359,6 +360,67 @@ ::util::Status BFChassisManager::VerifyChassisConfig(
     return MAKE_ERROR(ERR_INVALID_PARAM)
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:360]: CHECK_RETURN_IF_FALSE

Comment (bocon13) [reply stratum/hal/lib/barefoot/bf_chassis_manager.cc:360]: in the two above

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 5]
```diff
@@ -382,6 +444,13 @@ ::util::Status BFChassisManager::VerifyChassisConfig(
              << "newly pushed config requires a change in the port layout. "
              << "The stack needs to be rebooted to finish config push.";
     }
+
+    if (node_id_to_unit != node_id_to_unit_) {
+      return MAKE_ERROR(ERR_REBOOT_REQUIRED)
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:449]:
```suggestion
      RETURN_ERROR(ERR_REBOOT_REQUIRED)
```

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 6]
```diff
@@ -351,13 +352,87 @@ ::util::Status BFChassisManager::PushChassisConfig(
...
+  CHECK_RETURN_IF_FALSE(config.port_groups_size() == 0)
+      << "Port groups are not supported on Tofino.";
+  CHECK_RETURN_IF_FALSE(config.nodes_size() > 0)
+      << "The config must contain at least one node.";
+
+  // Find the supported Tofino chip types based on the given platform.
+  CHECK_RETURN_IF_FALSE(config.has_chassis() && config.chassis().platform())
+      << "Config needs a Chassis message with correct platform.";
+  CHECK_RETURN_IF_FALSE(config.chassis().platform() == PLT_BAREFOOT_TOFINO)
+      << "Unsupported platform: " << Platform_Name(config.chassis().platform());
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:366]: I'm surprised this was a) not checked, b) never tripped anyone up.

Comment (Yi-Tseng) [reply stratum/hal/lib/barefoot/bf_chassis_manager.cc:366]: I believed that we also need to add `PLT_BAREFOOT_TOFINO2`

Comment (pudelkoM) [reply stratum/hal/lib/barefoot/bf_chassis_manager.cc:366]: whoops

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 7]
```diff
@@ -351,13 +352,94 @@ ::util::Status BFChassisManager::PushChassisConfig(
...
+
+  // Find the supported Tofino chip types based on the given platform.
+  CHECK_RETURN_IF_FALSE(config.has_chassis() && config.chassis().platform())
+      << "Config needs a Chassis message with correct platform.";
+  switch (config.chassis().platform()) {
+    case PLT_BAREFOOT_TOFINO:
+    case PLT_BAREFOOT_TOFINO2:
+      break;
+    default:
+      return MAKE_ERROR(ERR_INTERNAL)
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:370]: ERR_INVALID_PARAM?

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 8]
```diff
@@ -351,13 +352,94 @@ ::util::Status BFChassisManager::PushChassisConfig(
...
+        << PrintSingletonPort(singleton_port)
+        << " was already recorded for another SingletonPort in the config.";
+    singleton_port_keys.insert(singleton_port_key);
+    CHECK_RETURN_IF_FALSE(singleton_port.node() > 0)
+        << "No valid node ID in " << singleton_port.ShortDebugString() << ".";
+    CHECK_RETURN_IF_FALSE(node_id_to_unit.count(singleton_port.node()))
+        << "Node ID " << singleton_port.node() << " given for SingletonPort "
+        << PrintSingletonPort(singleton_port)
+        << " has not been given to any Node in the config.";
+    CHECK_RETURN_IF_FALSE(
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:437]: Does this mean that port IDs only need to be unique for a given node? Said another way, this check verifies that (port id, node id) is unique...

Comment in ChassisConfig message says that the port ID needs to be unique across all ports (including across other port types i.e. trunks).

Comment (pudelkoM) [reply stratum/hal/lib/barefoot/bf_chassis_manager.cc:437]: > this check verifies that (port id, node id) is unique

This is correct. The tuple (node, port id) must be unique per switch.

The code is taken nearly verbatim from the BCM chassis manager, but we could ask at our next sync.

To me this seems sensible. Given that the port ID is a configurable SDN port, why should it not be allows to have port 1 on both forwarding planes? P4RT requests always contain the node ID.

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 9]
```diff
@@ -351,13 +352,94 @@ ::util::Status BFChassisManager::PushChassisConfig(
...
+  CHECK_RETURN_IF_FALSE(config.port_groups_size() == 0)
+      << "Port groups are not supported on Tofino.";
+  CHECK_RETURN_IF_FALSE(config.nodes_size() > 0)
+      << "The config must contain at least one node.";
+
+  // Find the supported Tofino chip types based on the given platform.
+  CHECK_RETURN_IF_FALSE(config.has_chassis() && config.chassis().platform())
+      << "Config needs a Chassis message with correct platform.";
+  switch (config.chassis().platform()) {
+    case PLT_BAREFOOT_TOFINO:
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:366]:
```suggestion
    case PLT_BAREFOOT_TOFINO:
      LOG(INFO) << "PLT_BAREFOOT_TOFINO is deprecated";
      // fall through
    case PLT_GENERIC_BAREFOOT_TOFINO:
```

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 10]
```diff
@@ -351,13 +352,94 @@ ::util::Status BFChassisManager::PushChassisConfig(
...
+    case PLT_BAREFOOT_TOFINO2:
+      break;
+    default:
+      return MAKE_ERROR(ERR_INVALID_PARAM)
+             << "Unsupported platform: "
+             << Platform_Name(config.chassis().platform());
+  }
+
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:374]: LOG a deprecation notice if PLT_BAREFOOT_TOFINO || PLT_BAREFOOT_TOFINO2

File: stratum/hal/lib/barefoot/bf_chassis_manager.cc [hunk 11]
```diff
@@ -351,13 +352,94 @@ ::util::Status BFChassisManager::PushChassisConfig(
...
+      << "Port groups are not supported on Tofino.";
+  CHECK_RETURN_IF_FALSE(config.nodes_size() > 0)
+      << "The config must contain at least one node.";
+
+  // Find the supported Tofino chip types based on the given platform.
+  CHECK_RETURN_IF_FALSE(config.has_chassis() && config.chassis().platform())
+      << "Config needs a Chassis message with correct platform.";
+  switch (config.chassis().platform()) {
+    case PLT_BAREFOOT_TOFINO:
+    case PLT_BAREFOOT_TOFINO2:
```
Comment (bocon13) [inline stratum/hal/lib/barefoot/bf_chassis_manager.cc:367]:
```suggestion
    case PLT_BAREFOOT_TOFINO:
    case PLT_BAREFOOT_TOFINO2:
    case PLT_GENERIC_BAREFOOT_TOFINO:
    case PLT_GENERIC_BAREFOOT_TOFINO2:
```

**Review by bocon13 — DISMISSED** [review]
lgtm

**Pull request merged**